src/tribol/mesh/MeshData.cpp

```diff
@@ -226,207 +194,277 @@ bool MeshElemData::isValidRatePenalty( PenaltyEnforcementOptions& pen_options )
+    }
   }
 
-  if (m_area != nullptr)
-  {
-    delete[] m_area;
-    m_area = nullptr;
-  }
+  // compute number of unique integer ids
+  int unique_size = sorted_conn.size() - num_dup;
 
-  if (m_sortedSurfaceNodeIds != nullptr)
+  SLIC_ERROR_IF(unique_size <= 0, "MeshData::sortSurfaceNodeIds(): " << 
+                "invalid connectivity array; " <<
+                "only single unique id in connectivity array.");
+
+  // allocate array to store unique, sorted node ids on mesh object
+  auto sorted_surface_node_ids = ArrayT<IndexT>(0, unique_size);
+
+  // populate sorted node id list
+  sorted_surface_node_ids.push_back(sorted_conn[0]);
+  for (IndexT i{1}; i < sorted_conn.size(); ++i)
   {
-     delete[] m_sortedSurfaceNodeIds;
-     m_sortedSurfaceNodeIds = nullptr;
+    if ( sorted_conn[i] != sorted_conn[i-1] )
+    { 
+      sorted_surface_node_ids.push_back(sorted_conn[i]);
+    }
   }
 
-}
+  return sorted_surface_node_ids;
+} // end MeshData::sortSurfaceNodeIds()
 
 //------------------------------------------------------------------------------
-bool MeshData::computeFaceData( int const dim )
+bool MeshData::computeFaceData()
 {
-  bool faceDataOk = true;
-  real fac = 1.0 / m_numNodesPerCell;
-  constexpr real nrmlMagTol = 1.0e-15;
-
-  // loop over all cells in the mesh
-  for (int i=0; i<m_numCells; ++i) {
-
-     // compute the vertex average centroid. This will lie in the 
-     // plane of the face for planar faces, and will be used as 
-     // an approximate centroid for warped faces, both in 3D.
-
-     // loop over the nodes per cell
-     for (int j=0; j<m_numNodesPerCell; ++j) {
-        auto nodeIndex = m_numNodesPerCell * i + j;
-        auto nodeId = m_connectivity[ nodeIndex ];
-        // always compute the x and y components for 2D and 3D
-        m_cX[i] += m_positionX[ nodeId ];
-        m_cY[i] += m_positionY[ nodeId ];
-     } // end loop over nodes
+  constexpr RealT nrml_mag_tol = 1.0e-15;
+
+  // allocate m_c
+  m_c = Array2D<RealT>({m_dim, numberOfElements()}, m_allocator_id);
+  // initialize m_c
+  m_c.fill(0.0);
+
+  // allocate m_n
+  m_n = Array2D<RealT>({m_dim, numberOfElements()}, m_allocator_id);
+  // initialize m_n
+  m_n.fill(0.0);
+
+  // allocate m_area (initialized to 0.0)
+  m_area = Array1D<RealT>(numberOfElements(), numberOfElements(), m_allocator_id);
+
+  // allocate face_radius (initialized to 0.0)
+  m_face_radius = Array1D<RealT>(numberOfElements(), numberOfElements(), m_allocator_id);
+  
+  ArrayT<IndexT> face_data_ok_data({static_cast<IndexT>(true)}, m_allocator_id);
+
+  // loop over all elements in the mesh
+  Array2DView<RealT> c = m_c;
+  MultiViewArrayView<const RealT> x = m_position;
+  Array2DView<RealT> n = m_n;
+  Array1DView<RealT> area = m_area;
+  Array1DView<RealT> radius = m_face_radius;
+  auto dim = m_dim;
+  auto conn = m_connectivity;
+  ArrayViewT<IndexT> face_data_ok = face_data_ok_data;
+  forAllExec(getExecutionMode(m_mem_space), numberOfElements(), 
```

Comment:
I think we talked about computeFaceData() taking in an execution mode and not using `getExecutionMode(m_mem_space)`. Please verify.